### ISSUE TITLE
Promotion rules for same base color pairs

### ISSUE THREAD

**Issue by kimikage** [JuliaGraphics/ColorTypes.jl#129]
As https://github.com/JuliaGraphics/Colors.jl/issues/273 shows, there are the problems with promotion rules for same base color pairs. Here is another simple example:
```julia
julia> promote_type(HSV{Float64}, HSV{Float32})
HSV

julia> promote(HSV{Float64}(0,0,0), HSV{Float32}(0,0,0))
ERROR: promotion of types HSV{Float64} and HSV{Float32} failed to change any arguments
Stacktrace:
 [1] error(::String, ::String, ::String) at .\error.jl:42
 [2] sametype_error(::Tuple{HSV{Float64},HSV{Float32}}) at .\promotion.jl:308
 [3] not_sametype(::Tuple{HSV{Float64},HSV{Float32}}, ::Tuple{HSV{Float64},HSV{Float32}}) at .\promotion.jl:302
 [4] promote(::HSV{Float64}, ::HSV{Float32}) at .\promotion.jl:285
 [5] top-level scope at REPL[2]:1
```

A simple (and tactless) solution is adding the rules [here](https://github.com/JuliaGraphics/ColorTypes.jl/blob/7ef745605d6a0897078712a0a7a156c51295b559/src/types.jl#L549-L579).
```julia
    @eval Base.promote_rule(::Type{$Csym{T1}}, ::Type{$Csym{T2}}) where {T1, T2} = $Csym{promote_type(T1, T2)}
    @eval Base.promote_rule(::Type{$acol{T1}}, ::Type{$acol{T2}}) where {T1, T2} = $acol{promote_type(T1, T2)}
    @eval Base.promote_rule(::Type{$cola{T1}}, ::Type{$cola{T2}}) where {T1, T2} = $cola{promote_type(T1, T2)}
```
Of course,
> RGB1 and RGB4 require special handling because of the alphadummy field

However, I don't know what should be returned by `promote_type(HSV, HSV{Float16})`.(`HSV{Float16}`? or `HSV{promote_type(eltype_default(HSV),Float16)}`?)

**Comment by kimikage** [comment]
Recently, I am thinking that I should leave it to the compiler optimization and write `if` statements without `@generated`.

**Comment by kimikage** [comment]
The promotion rules introduced in PR #147 will not see the light of day.:sweat_smile: 
```julia
function Base.promote_rule(::Type{T1}, ::Type{T2}) where {T1<:Colorant, T2<:Colorant}
    et = promote_type(eltype(T1), eltype(T2))
    base_colorant_type(T1) === base_colorant_type(T2) && return base_colorant_type(T1){et}
    T1 === ARGB32 && T2 <: AbstractARGB && return ARGB{et}
    T2 === ARGB32 && T1 <: AbstractARGB && return ARGB{et}
    T1 === AGray32 && T2 <: AbstractAGray && return AGray{et}
    T2 === AGray32 && T1 <: AbstractAGray && return AGray{et}
    T1 <: AbstractGray && return base_colorant_type(T2){et}
    T2 <: AbstractGray && return base_colorant_type(T1){et}
    T1 <: TransparentGray && return base_colorant_type(T2){et}
    T2 <: TransparentGray && return base_colorant_type(T1){et}
    ccolor(T1, T2)
end
function Base.promote_rule(::Type{T1}, ::Type{T2}) where {T1<:TransparentColor, T2<:Color}
    et = promote_type(eltype(T1), eltype(T2))
    base_color_type(T1) === base_colorant_type(T2) && return base_colorant_type(T1){et}
    T1 <: AbstractAGray && return alphacolor(base_color_type(T2)){et}
    T1 <: AbstractGrayA && return coloralpha(base_color_type(T2)){et}
    T2 <: AbstractGray  && return base_colorant_type(T1){et}
    ccolor(T1, T2)
end
function Base.promote_rule(::Type{T1}, ::Type{T2}) where {T1<:AbstractRGB, T2<:AbstractRGB}
    et = promote_type(eltype(T1), eltype(T2))
    base_colorant_type(T1) === base_colorant_type(T2) && return base_colorant_type(T1){et}
    T1 <: RGBX && !(T2 <: XRGB) && return RGBX{et}
    T1 <: XRGB && !(T2 <: RGBX) && return XRGB{et}
    T2 <: RGBX && !(T1 <: XRGB) && return RGBX{et}
    T2 <: XRGB && !(T1 <: RGBX) && return XRGB{et}
    RGB{et}
end
function Base.promote_rule(::Type{T1}, ::Type{T2}) where {T1<:TransparentRGB, T2<:AbstractRGB}
    et = promote_type(eltype(T1), eltype(T2))
    T1 === ARGB32 && T2 === RGB24 && return ARGB32
    T1 === ARGB32 && return ARGB{et}
    base_colorant_type(T1){et}
end
function Base.promote_rule(::Type{T1}, ::Type{T2}) where {T1<:AbstractGray, T2<:AbstractGray}
    et = promote_type(eltype(T1), eltype(T2))
    base_colorant_type(T1) === base_colorant_type(T2) && return base_colorant_type(T1){et}
    Gray{et}
end
function Base.promote_rule(::Type{T1}, ::Type{T2}) where {T1<:TransparentGray, T2<:AbstractGray}
    et = promote_type(eltype(T1), eltype(T2))
    T1 === AGray32 && T2 === Gray24 && return AGray32
    T1 === AGray32 && return AGray{et}
    base_colorant_type(T1){et}
end

function Base.promote_rule(::Type{T1}, ::Type{T2}) where {T1<:Color, T2<:TransparentColor}
    promote_rule(T2, T1)
end
```
We need additional rules and tests for non-concrete types.

**Comment by kimikage** [comment]
Oops! :scream:
https://github.com/JuliaGraphics/ColorTypes.jl/blob/5b2e50d62b8fd47a860c2e0543446144d8627c87/test/conversions.jl#L102-L107

**Comment by kimikage** [comment]
However, my concern is that even though `promote_rule` and `ccolor` have almost the same functionality, the codes are different now.
The difference between the two is that `ccolor` has the order of priority between its arguments.

**Comment by timholy** [comment]
`promote_rule` and `ccolor` really are different from one another: promotion is about "type-widening," i.e., common up with a type that can store both values. `ccolor` is closer to "type-narrowing," e.g., coming up with a more specific type from general guidance.

**Comment by kimikage** [comment]
That's right. However, promotion rules doesn't have to match the type hierarchy. In the `ColorTypes` hierarchy, we can too easily get `Any`. That's reasonable, but not really useful. If we want to get a purely common type, we should use `typeintersect`.
However, at the moment, only limited promotions are working, so the useless rules may be fine.

**Comment by timholy** [comment]
`typejoin` is closer to what we want than `typeintersect`, but you're right that what I wrote earlier is more appropriate for `typejoin` than `promote_type`. (And indeed that's what it already gives. We should not specialize `typejoin` since that's a generic reflection tool.)

If there's a type that includes the gamut of most/all other types (`XYZ`?), we could promote to that. A little loss in precision is acceptable, just as with `promote_type(Int, Float64) === Float64`.

**Comment by kimikage** [comment]
I think it's more convenient to allow the user to specify (i.e. override) the rules for concrete types such as `RGB{T}` vs `HSV{T}`. 

I will change the unclear rules to the rules purely based on the type hierarchy.

**Comment by timholy** [comment]
Yes, I do think `typejoin`-like behavior is a reasonable fallback.

**Comment by kimikage** [comment]
To be honest, I do not have a clear answer on what the promotion rules for non-concrete types should be. 
`promote_type(AbstractRGB{Float32}, RGB24)` returns `RGB{Float32}` for now, but is that OK? (cf. https://github.com/JuliaGraphics/ColorTypes.jl/pull/161#discussion_r378930542)
I believe that `RGB{Float32}` is more pragmatic than `AbstractRGB{Float32}`. However I doubt that this is the exceptional rule allowed only for `AbstractRGB` and `AbstractGray`. In particular, `AbstractGray` is a typealias, and its substance is `Color`.

**Comment by timholy** [comment]
> To be honest, I do not have a clear answer on what the promotion rules for non-concrete types should be.

Yes, this is difficult terrain. I think the best guidance is to think about allocating a container of concrete `eltype` that could store objects of all the specified types reasonably losslessly (whatever that means :man_shrugging: ). If that requires conversion, that's OK.

> promote_type(AbstractRGB{Float32}, RGB24) returns RGB{Float32}

Yes, that seems clearly correct.

I think I'd be pretty happy with the following rules (but I doubt they are sufficient), which I do think you've basically implemented:

Colors:
- `C{T1}` and `C{T2}` promote to `C{T}`
- `C1{T1}` and `C2{T2}` promote to `RGB{T}` if `C1<:AbstractRGB`, `C2<:AbstractRGB` 
- `C1{T1}` and `C2{T2}` promote to `Gray{T}` if `C1<:AbstractGray`, `C2<:AbstractGray` 
- if one is `<:AbstractGray` and the other `<:AbstractRGB` promote to whatever RGB-type the one has (if it's concrete); if it's just plain `AbstractRGB` then pick `RGB`. 
- otherwise, anything with 3 or fewer color channels promotes to `XYZ{T}`

Transparency:
- if both `C1` and `C2` are under the same `AlphaColor` or `ColorAlpha`, use that
- if only one has transparency, and it's either `AlphaColor` or `ColorAlpha`, use that
- prefer `AlphaColor` otherwise

One thing I've probably never mentioned: part of the design of this package was to support custom color spaces (it's why a lot of the constructor-generators are macros). To give a concrete example of where that's useful, a lot of biological imaging is done with two color channels, for example [Green Fluorescent Protein](https://en.wikipedia.org/wiki/Green_fluorescent_protein) (see [spectrum](https://www.fpbase.org/protein/egfp/)) and [tdTomato](https://www.fpbase.org/protein/tdtomato/). If you've collected data from both color channels over the same spatial region, ideally the resulting array would have eltype `GFPtdTomato{Float32}` or something, where

```
struct GFPtdTomato{T} <: Color{T,2}
    gfp::T
    tdTomato::T
end
```

That, of course, introduces a whole new class of headaches wrt promotion rules. You don't have to handle that, but that's part of why I said "3 or fewer color channels."

**Comment by kimikage** [comment]
Oh, I changed the rules to the `typejoin`-like before you posted the comment above.
```julia
julia> using ColorTypes

julia> promote_type(RGB, RGB{Float16})
RGB

julia> promote_type(HSV, HSV{Float16})
HSV

julia> promote_type(AbstractRGB, RGB{Float16})
RGB

julia> promote_type(AbstractRGB{Float32}, RGB{Float16})
RGB{Float32}
```

> if one is `<:AbstractGray` and the other `<:AbstractRGB` promote to whatever RGB-type the one has (if it's concrete); if it's just plain `AbstractRGB` then pick `RGB`.

In addition, at present (i.e. Colors.jl v0.11), it does not matter what the gray opponent is.
```julia
julia> promote_type(HSV{Float32},Gray{Float64})
HSV{Float64}
```
Of course, there is a difference that RGB-type has `AbstractRGB` and HSV-type does not have `AbstractHSV`.
```julia
julia> promote_type(Color3{Float32},Gray{Float64}) # Colors v0.11
Color{Float64,N} where N
```
```julia
julia> promote_type(Color3{Float32},Gray{Float64}) # PR #164
Color{Float64,3} where 3
```

**Comment by kimikage** [comment]
> otherwise, anything with 3 or fewer color channels promotes to `XYZ{T}`

This is a rule I (we) have not implemented. I agree with this rule if we want to force promotion to concrete types, but if not, I don't really like the rule.

Although ColorTypes.jl is not involved in the color space conversions, Colors.jl actually converts the color spaces as follows. (cf. https://github.com/JuliaGraphics/Colors.jl/issues/354#issuecomment-534547652)
![65509304-22078d80-df0d-11e9-8b98-66cbaca3c740](https://user-images.githubusercontent.com/12679384/74794694-35c1ef00-5307-11ea-9951-8a76650be5ea.png)
Except for the direct conversion system that uses a lookup table, the most conversion flow draws such a graph (with some little differences).
The important point is that in addition to `XYZ`, there are three hubs: `RGB`, `Lab`, and `Luv`.

For example, it is good for consistency that the promoted type of `Lab` and `LCHab` is `XYZ`, but it doesn't seem really practical. However, the semantic relationship between `Lab` and `LCHab` is clear, but the semantic relationship between `HSV` and `RGB` is slightly complicated (https://en.wikipedia.org/wiki/HSL_and_HSV#/media/File:HSL-HSV_hue_and_chroma.svg).
I don't think it is really good to go into such a detail of color spaces.

Another concern for me is that `XYZ` is an absolute color space, but other color spaces may be relative. The sRGB is an absolute color space, and there is no ambiguity in conversion to `XYZ`. Although README of ColorTypes.jl describes the `RGB` means the sRGB, the rules are not strictly followed in actual image processing. I also think that sRGB is "too narrow" for some applications.

On the other hand, even though `Lab` (CIELAB) is an absolute color space, it depends on the light source or "white point", so it is somewhat relative.

Of course, we can use the default context (e.g. gamut or white point), but the users cannot control the context  within Colors.jl.

**Comment by kimikage** [comment]
Based on the issues above, my strategy was to hide the rules except for trivial rules for `RGB24` and `RGB32` and so on.
```julia
Base.promote_rule(::Type{C1}, ::Type{C2}) where {C1<:Colorant, C2<:Colorant} = _promote_rule(C1, C2)
```
This makes it difficult for users to the directly intervene in the rules, but this allows to easily override the rules. (Of course, being able to write rules easily is quite different from that the rules are easy, though. :innocent:)

**Comment by kimikage** [comment]
> prefer AlphaColor otherwise

I *did* not agree to this rule because `AlphaColor` has no legitimacy to take precedence over `ColorAlpha`. In fact, CSS uses `RGBA` instead of `ARGB`. Therefore, the rules implemented in PR #164 promote `AlphaColor` and `ColorAlpha` to `TransparentColor`.

However, these generalizations (i.e. uplifting in the type hierarchy) make the rules too complicated, mainly due to the irregular rules related to `AbstractRGB`.

TBH, until two days ago, I believed that the rules could be implemented with ad-hoc and ugly codes. However, when I simulated adding the premultiplied alpha types for issue #167, I realized that the ad-hoc codes have no scalability. To be precise, I knew they were not scalable, but I thought that the only possible change in the type hierarchy was the addition of concrete color types (e.g. [HWB](https://www.w3.org/TR/2019/WD-css-color-4-20191105/#the-hwb-notation)).

Thus, I realized the practical need to bias either `AlphaColor` or `ColorAlpha`.
I'm writing the rules which prioritizes `AlphaColor`. However, as mentioned above, I'm going not to include the promotions to `XYZ` in the rules. The promotions to `Color`/`ColorAlpha`/`AlphaColor` do not complicate the rules so much, since it concerns the lowest layer of the hierarchy.
